eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs

```diff
@@ -909,9 +910,13 @@ consumed ::
   UTxO era ->
   Core.TxBody era ->
   Core.Value era
-consumed pp u tx =
-  balance @era (eval (txins @era tx ◁ u)) <> Val.inject (refunds <+> withdrawals)
+consumed pp (UTxO u) tx =
+  {- balance (txins tx ◁ u) + wbalance (txwdrls tx) + keyRefunds pp tx -}
+  Set.foldl' lookupAddTxOut mempty (txins @era tx)
+    <> Val.inject (refunds <+> withdrawals)
   where
+    lookupAddTxOut acc txin = maybe acc (addTxOut acc) $ SplitMap.lookup txin u
+    addTxOut !b out = getField @"value" out <+> b
```

Comment:
Perhaps like this?

consumed pp u tx =
  balance @era (txins @era tx Split.◁ u) <> Val.inject (refunds <+> withdrawals)

Reply:
This operation `txins @era tx Split.◁ u)` will create an intermediate `Set`, while doing a fold on `txins` avoids that

SplitMap will already introduce a regression in performance, we need to hold on to any saving we can get.